Add Proxy Transport service

rosstimothy opened the pull request:
The first step in implementing the `teleport.proxy.v1.ProxyService` as defined in RFD 100. This includes part of the service implementation and a stream abstraction to allow transporting payloads for frames over a gRPC stream. Inspiration for the stream abstraction was taken from [proxy peering](https://github.com/gravitational/teleport/blob/master/lib/proxy/peer/conn.go).

Part of #19812

zmb3 commented on lib/srv/transport/stream.go:
```diff
@@ -0,0 +1,135 @@
+
+// Stream abstracts gRPC streams used to transport
+// frames of data.
+type Stream interface {
```
I realize this is based off the gRPC interface, but it looks awfully similar to `io.Reader` and `io.Writer` - have you thought about reusing those interfaces here too?

Notably, the signature of `Receive` lends itself to doing lots of allocations rather than writing into an existing buffer.

zmb3 commented on lib/srv/transport/stream.go:
```diff
@@ -0,0 +1,135 @@
+	s.rLock.Lock()
+	defer s.rLock.Unlock()
+
+	// return any buffered data prior to
+	// pulling from the stream again
```
nit: the comment is kind of describing the inverse of what the code is doing.

maybe rephrase it?

```
// read from the stream only if we don't have 
// data leftover from previous calls
```

zmb3 commented on lib/srv/transport/stream.go:
```diff
@@ -0,0 +1,135 @@
+	// maxChunkSize is the maximum number of bytes to send in a single data message.
+	// According to https://github.com/grpc/grpc.github.io/issues/371 the optimal
+	// size is between 16KiB to 64KiB.
+	maxChunkSize int = 1024 * 16
```
If the optimal size is between 16 and 64, and you set the max size to 16, aren't you kind of guaranteeing that most chunks will be suboptimal?

zmb3 commented on lib/srv/transport/stream.go:
```diff
@@ -0,0 +1,135 @@
+		}
+
+		if err := s.stream.Send(chunk); err != nil {
+			return 0, trace.ConnectionProblem(err, "failed to send on frame")
```
If p is large, and this fails after the first write, then you're violating the io.Writer contract by reporting that 0 bytes were written.

zmb3 commented on lib/srv/transport/v1/transport.go:
```diff
@@ -0,0 +1,120 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
```
What's the reasoning behind this package name?

zmb3 commented on lib/srv/transport/v1/transport.go:
```diff
@@ -0,0 +1,120 @@
+)
+
+// Router is the interface that groups basic dialing methods.
+type Router interface {
```
We already have 2 `Router` types in Teleport. I'm afraid that introducing any more will make it really hard to tell when to use each of them.

tobiaszheller commented on lib/srv/transport/transportv1/transport.go:
```diff
@@ -0,0 +1,120 @@
+// Dialer is the interface that groups basic dialing methods.
+type Dialer interface {
+	DialSite(ctx context.Context, clusterName string) (net.Conn, error)
+	DialHost(ctx context.Context, from net.Addr, host, port, clusterName string, accessChecker services.AccessChecker, agentGetter teleagent.Getter) (net.Conn, error)
```
that will be used in other PR right?

tobiaszheller commented on lib/srv/transport/transportv1/transport.go:
```diff
@@ -0,0 +1,120 @@
+
+// Dialer is the interface that groups basic dialing methods.
+type Dialer interface {
+	DialSite(ctx context.Context, clusterName string) (net.Conn, error)
```
why is it called dialSite? I saw that naming already in teleport codebase but never understand why it's named site

tobiaszheller commented on lib/srv/transport/transportv1/transport.go:
```diff
@@ -0,0 +1,120 @@
+		return trace.Wrap(err, "failed dialing cluster %q", req.Cluster)
+	}
+
+	if req.Frame != nil {
```
what we should do if frame is nil? don't we require it for the first time? 
Imo it would be good to add some docs where why we are doing writes for first time

tobiaszheller reviewed (commented):
good job at reusing `utils.ProxyConn`, left few questions

zmb3 commented on lib/utils/grpc/stream.go:
```diff
@@ -0,0 +1,122 @@
+	rBytes []byte
+}
+
+// NewStreamer creates a new streamConn.
```
`streamConn`?

zmb3 commented on lib/utils/grpc/stream.go:
```diff
@@ -0,0 +1,122 @@
+	if len(c.rBytes) == 0 {
+		data, err := c.stream.Recv()
+		if errors.Is(err, io.EOF) {
+			return 0, io.EOF
```
Is it possible to get an `io.EOF` and valid data in the same call?

tobiaszheller commented on lib/proxy/peer/client.go:
```diff
@@ -356,7 +357,41 @@ func (c *Client) DialNode(
+	return newStreamConn(streamRW, src, dst), nil
+}
+
+// source is the common subset of the [clientapi.ProxyService_DialNodeClient] and
```
this comment if off, it should describe `stream` interface

tobiaszheller reviewed (approved):
lgtm, left 1 comment

rosstimothy committed:
Add Proxy Transport service

The first step in implementing the teleport.proxy.v1.ProxyServic
as defined in RFD 100. This includes part of the service
implementation and a stream abstraction to allow transporting
payloads for frames over a gRPC stream.

The proxy peering specific implementation of the stream abstraction
was refactored into `lib/utils/grpc/stream` so that other services
can make use of it.

Part of #19812
